src/main/java/org/prebid/server/gdpr/vendorlist/VendorListService.java

```diff
@@ -0,0 +1,253 @@
+        }
+        return cache;
+    }
+
+    /**
+     * Reads files with .json extension in configured directory and
+     * returns a {@link Map} where key is a file name without .json extension and value is file content.
+     */
+    private static Map<String, String> readFileSystemCache(FileSystem fileSystem, String dir) {
+        return fileSystem.readDirBlocking(dir).stream()
+                .filter(filepath -> filepath.endsWith(JSON_SUFFIX))
+                .collect(Collectors.toMap(filepath -> StringUtils.removeEnd(new File(filepath).getName(), JSON_SUFFIX),
+                        filename -> fileSystem.readFileBlocking(filename).toString()));
+    }
+
+    /**
+     * Creates {@link VendorListInfo} object from string content or throw {@link PreBidException}.
+     */
+    private static VendorListInfo toVendorListInfo(String content) {
+        try {
+            return Json.mapper.readValue(content, VendorListInfo.class);
+        } catch (IOException e) {
+            final String message = String.format("Cannot parse vendor list from: %s", content);
+
+            logger.error(message, e);
+            throw new PreBidException(message, e);
+        }
+    }
+
+    /**
+     * Returns a map with vendor ID as a key and set of purposes as a value from given list of {@link Vendor}s.
+     */
+    private static Map<Integer, Set<Integer>> mapVendorIdToPurposes(List<Vendor> vendors, Set<Integer> knownVendorIds) {
+        return vendors.stream()
+                .filter(vendor -> knownVendorIds.contains(vendor.getId())) // optimize cache to use only known vendors
+                .collect(Collectors.toMap(Vendor::getId, Vendor::getPurposeIds));
+    }
+
+    /**
+     * Returns a map with vendor ID as a key and a set of purposes as a value for given vendor list version.
+     */
+    public Future<Map<Integer, Set<Integer>>> forVersion(int version) {
+        final Map<Integer, Set<Integer>> vendorIdToPurposes = cache.get(version);
+        if (vendorIdToPurposes != null) {
+            return Future.succeededFuture(vendorIdToPurposes);
+        } else {
+            fetchNewVendorListFor(version);
+
+            return Future.failedFuture(String.format("Vendor list of version %d not found. Try again later.", version));
+        }
+    }
+
+    /**
+     * Proceeds obtaining new vendor list from HTTP resource.
+     */
+    private void fetchNewVendorListFor(int version) {
+        httpClient.getAbs(String.format(endpointTemplate, version),
+                response -> handleResponse(response, version))
+                .exceptionHandler(throwable -> handleException(version, throwable))
+                .setTimeout(defaultTimeoutMs)
+                .end();
+    }
+
+    private static void handleException(int version, Throwable throwable) {
+        fail(version, throwable.getMessage());
+    }
+
+    private static void fail(int version, String errorMessageFormat, Object... args) {
```

Comment:
Since this method just logs error it could be named accordingly, for example `logFetchError`